R/germline.R

```diff
@@ -61,51 +60,75 @@ repGermline <- function(.data, species = "HomoSapiens") {
   }
 }
 
-germline_single_df <- function(data, species) {
-  # add first allele of V and J genes
-  data["V.first.allele"] <- data %>%
-    select(V.name) %>%
-    apply(MARGIN = 1, FUN = take_first_allele)
-  data["J.first.allele"] <- data %>%
-    select(J.name) %>%
-    apply(MARGIN = 1, FUN = take_first_allele)
-
-  # load V genes
-  v_genes <- load_reference_sequences("IGHV", species)
-
-  # add V genes
-  data <- merge(x = v_genes, y = data, by = "V.first.allele", all.y = TRUE)
-
-  # load J genes
-  j_genes <- load_reference_sequences("IGHJ", species)
-
-  # add J genes
-  data <- merge(x = j_genes, y = data, by = "J.first.allele", all.y = TRUE)
-
+germline_single_df <- function(data, species, sample_name = NA) {
   data %>%
-    mutate(Germline.Sequence = purrr::map2(
-      V.sequence, J.sequence, generate_germline_sequence
-    )) %>%
-    return()
+    rowwise() %>%
+    mutate(V.first.allele = take_first_allele(V.name)) %>%
+    merge_reference_sequences("V", species, sample_name) %>%
+    rowwise() %>%
+    mutate(Germline.sequence = generate_germline_sequence(
+      Sequence, V.sequence, V.end, CDR3.start, CDR3.end, J.start
+    ))
 }
 
 take_first_allele <- function(string) {
   unlist(strsplit(string, ","))[1]
 }
 
-generate_germline_sequence <- function(v_seq, j_seq) {
-  if (is.na(v_seq) || is.na(j_seq)) {
+generate_germline_sequence <- function(seq, v_ref, v_end, cdr3_start, cdr3_end, j_start) {
+  if (any(sapply(c(seq, v_ref, v_end, cdr3_start, cdr3_end, j_start), is.na))) {
     return(NA)
   } else {
-    return(paste(v_seq, j_seq, sep = "..."))
+    cdr3_start %<>% as.numeric()
+    cdr3_end %<>% as.numeric()
```

Comment:
They should already be in numeric format when immunarch class is created

Reply:
In fact, it's not: it is a string, and it contains original unparsed MiXCR coordinate string as it's name. Maybe, it's a bug in MiXCR parser in repLoad().

Reply:
Please check this. @vadimnazarov mentioned that repLoad will produce data in immunarch format, making these lines unnecessary

Reply:
I've checked: removing of these lines causes this function to crash because of numeric operations with non-numeric values